### DISCUSSION TITLE
v5.0.4 Developer Preview

### DISCUSSION THREAD

**Discussion by xiaokangwang** [v2fly/v2ray-core#1761] [General]
v5.0.4 Developer Preview 

This is a pre-release version of V2Ray intended for developer preview. Users are not advised to use this version right now. Contributors are encouraged to send pull requests to the V5 version of the codebase.
This release note is being edited to reflect all significant changes made to the codebase, please contact us if you think something is missing.
Most new features are currently undocumented. The documents will be updated over time.

## New Features
* **Jsonv5** New configuration format. This enables support for automatic component recognition. 
    * [ZH Document](https://www.v2fly.org/v5/config/overview.html) is work in progress.
    * [EN Document](https://www.v2fly.org/en_US/v5/config/overview.html)  is work in progress.
    * Thanks @AkinoKaede, @kslr, @mzz2017.
* **VLite** UDP P2P focused proxy protocol with Full Cone, Forward Error Correction, Self-Healing Connection(Connection Stabilization) support.
    *  UDP based VLite support was added in #1732.
    * The [ZH](https://www.v2fly.org/v5/config/proxy/vlite.html) [EN](https://www.v2fly.org/en_US/v5/config/proxy/vlite.html) documents have been updated.
* **UDP PacketAddr** UDP endpoint independent mapping(aka Full Cone) support. 
* **SocksOpt** Add KeepAlive Support. Thanks @ValdikSS .
* **SocksOpt** Add BindToDevice Support. Thanks @database64128 .
* **SocksOpt** Add Rx/TxBuf Size Support.
* **Burst Observatory** Measure connection quality to a significant amount of outbound at the same time. Thanks @qjebbs.
* **Multi Observatory** Run more than one Observatory at the same time.
* **Router** Load balancing strategy LeastLoad. Thanks @qjebbs.
* **Jsonv4** Additional representation format for Jsonv4 configure file format. This includes YAML, TOML. Thanks @qjebbs.
* **Jsonv4** General purpose merger for all configure formats. Thanks @qjebbs.
* **Shadowsocks** Reduced IV Head Entropy Experiment. Thanks GFWReport and other collaborators.
* **CLI** Reworked command line Interface.  Thanks @qjebbs.
* **Developer Assistance** Protobuf configure format reverse engineering.
* **Asset** Search for assets in xdg data directories on non-windows platforms #1578 Thanks @NickCao.
*  **DNS** Feat: refine find IPs logic for DoH. Thanks @AkinoKaede .
* **gRPC** Feat: make gRPC dialer accept socket config #1697 Thanks @dyhkwong

## Fixes
* **Router** Fixed a memory leak bug caused by requiring component.
*  **FakeDNS** Fix a concurrency issue in fakedns #1666 Thanks @yuhan6665 .
* **Jsonv5** Fix: context of DNS simplified config #1665 Thanks @AkinoKaede .
* **DNS** Fix: fix failure of cache expired DNS record querying over DoH #1706 Thanks @KujouRinka.
* **DNS** Fix: failure of cache expired DNS record querying over DoQ. Thanks @AkinoKaede .
* Fix client dialer log #1568 Thanks @yuhan6665 @AkinoKaede @kslr.
* Fix: net.Address type checking #1629 Thanks @1ocalhost

## Chores

* Test: Add statistics in conn testing #1444 Thanks @yuhan6665 @kslr .
* Chore: bump github.com/lucas-clemente/quic-go to 0.27.0 Thanks @AkinoKaede .
* Fix: lint setting for revive #1758 Thanks @Loyalsoldier 
* Fix: gci command for formatting code #1757 Thanks @Loyalsoldier 

<hr /><em>This discussion was created from the release <a href='https://github.com/v2fly/v2ray-core/releases/tag/v5.0.4'>v5.0.4 Developer Preview </a>.</em>

**Comment by PoneyClairDeLune** [comment]
Will VLite support forwarding TCP connections as well?
